src/Language/PureScript/Ide/Filter.hs

```diff
@@ -15,112 +15,130 @@
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
 
 module Language.PureScript.Ide.Filter
-       ( Filter
-       , declarationTypeFilter
-       , namespaceFilter
+       ( Filter(..)
        , moduleFilter
+       , namespaceFilter
+       , exactFilter
        , prefixFilter
-       , equalityFilter
+       , declarationTypeFilter
        , applyFilters
        ) where
 
-import           Protolude                     hiding (isPrefixOf)
+import           Protolude                     hiding (isPrefixOf, Prefix)
 
+import           Data.Bifunctor (first)
 import           Data.Aeson
-import           Data.List.NonEmpty            (NonEmpty)
 import           Data.Text                     (isPrefixOf)
-import qualified Language.PureScript.Ide.Filter.Declaration as D
+import qualified Data.Set as Set
+import qualified Data.Map as Map
+import           Language.PureScript.Ide.Filter.Declaration (DeclarationType, declarationType)
 import           Language.PureScript.Ide.Types
 import           Language.PureScript.Ide.Util
 import qualified Language.PureScript           as P
 
-newtype Filter = Filter (Endo [Module])
-  deriving (Semigroup, Monoid)
+newtype Filter = Filter (Either (Set P.ModuleName) DeclarationFilter)
```

Comment:
Have you considered having `Filter` be a `newtype` around `[Either (Set P.ModuleName) DeclarationFilter]`? Then I think you could still have a monoid instance which does something similar to `optimizeFilters`. Just a suggestion; it might be nice for consumers of this API not to have to worry about the difference between a single filter and a collection of filters.

Reply:
I did, I just wanted to keep the model closer to the protocol here. I'm thinking about making a new command which makes more assumptions about how editors end up asking for completions which could change things around a bit more.